### PR TITLE
bluestore/NVMEDevice.cc: fix NVMEManager thread hang

### PR DESCRIPTION
**Pull request by tone-zhang**
[bluestore/NVMEDevice]: [fix NVMEManager thread halt]

When enable SPDK in Ceph and start up Ceph development cluster, met NVMEManager thread halt.

On aarch64 platform, the log as below:

Starting SPDK v18.04.1 / DPDK 18.05.0 initialization...
[ DPDK EAL parameters: nvme-device-manager -c 0x1 -m 2048 --file-prefix=spdk_pid16987 ]
EAL: Detected 46 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/spdk_pid16987/mp_socket
EAL: Probing VFIO support...
EAL: VFIO support initialized
EAL: PCI device 0000:01:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
EAL:   using IOMMU type 1 (Type 1)
^C

The reason is that pthread_cond_destroy() cannot destroy the active condition_variable parameter.

Also on x86 debug builds we get the following error messages due to probe_queue_lock still being active during ~NVMEManager:

/home/ubuntu/ceph/src/common/mutex_debug.h: 114: FAILED ceph_assert(r == 0)
 ceph version 14.0.1-1862-g403622b (403622be721a460f3dff2d84f6bfc628f5026704) nautilus (dev)

The change fixes the issue.

Fixes: http://tracker.ceph.com/issues/37720

Signed-off-by: tone.zhang <tone.zhang@arm.com>
Signed-off-by: Steve Capper <steve.capper@arm.com>

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/os/bluestore/NVMEDevice.cc [hunk 1]
```diff
@@ -590,7 +602,7 @@ int NVMEManager::try_get(const spdk_nvme_transport_id& trid, SharedDriverData **
           spdk_nvme_retry_count = SPDK_NVME_DEFAULT_RETRY_COUNT;
 
         std::unique_lock l(probe_queue_lock);
-        while (true) {
+        while (should_probe {
```
Comment (tchaikov) [inline src/os/bluestore/NVMEDevice.cc]: this does not compile.

Comment (tone-zhang) [reply src/os/bluestore/NVMEDevice.cc]: Thanks. Will update soon.

File: src/os/bluestore/NVMEDevice.cc [hunk 2]
```diff
@@ -486,6 +487,17 @@ class NVMEManager {
 
  public:
   NVMEManager() {}
+  ~NVMEManager() {
+    if (!init)
```
Comment (tchaikov) [inline src/os/bluestore/NVMEDevice.cc]: we still need to wake up the `try_get()` waiters, so i think we should 

1. remove the member variable of `init`, and instead check if the `dpdk_thread` is joinable, before executing `dpdk_thread`, and join it in the dtor only if it's joinable.
1. set all pending `ctx->done` to `true`, before notifying all `try_get()` waiters in `dpdk_thread`.
1. might want to rename `should_probe` to `stopping`.

Comment (tone-zhang) [reply src/os/bluestore/NVMEDevice.cc]: Thanks Kefu.
For point 2, I will promote ctx as NVMEManager member variable then the destroyer can access it.

I will update the change soon. Thanks a lot!

Comment (tchaikov) [reply src/os/bluestore/NVMEDevice.cc]: i don't follow you. `ctx`s are elements of `probe_queue`. so i think we already have access to them? also, i'd recommend set them in `dpdk_thread` before it exits. it's more readable this way, as we notify the waiter in a single place.

Comment (tone-zhang) [reply src/os/bluestore/NVMEDevice.cc]: Got it. I mentioned another ctx. I will update code.

Thanks a lot! :)

File: src/os/bluestore/NVMEDevice.cc [hunk 3]
```diff
@@ -477,7 +477,7 @@ class NVMEManager {
 
  private:
   ceph::mutex lock = ceph::make_mutex("NVMEManager::lock");
-  bool init = false;
+  std::atomic<bool> stopping = true;
```
[the remark this answers is not in the document]
Comment (tone-zhang) [reply src/os/bluestore/NVMEDevice.cc]: It is more reasonable. Thanks. Will update the following code.

File: src/os/bluestore/NVMEDevice.cc [hunk 4]
```diff
@@ -600,14 +609,15 @@ int NVMEManager::try_get(const spdk_nvme_transport_id& trid, SharedDriverData **
               derr << __func__ << " device probe nvme failed" << dendl;
             }
             ctxt->done = true;
+            for (auto p : probe_queue)
+              p->done = true;
```
[the remark this answers is not in the document]
Comment (tone-zhang) [reply src/os/bluestore/NVMEDevice.cc]: Thanks. Will update the code soon.

File: src/os/bluestore/NVMEDevice.cc [hunk 5]
```diff
@@ -605,9 +614,10 @@ int NVMEManager::try_get(const spdk_nvme_transport_id& trid, SharedDriverData **
             probe_queue_cond.wait(l);
           }
         }
+        for (auto p : probe_queue)
+          p->done = true;
```
Comment (tchaikov) [inline src/os/bluestore/NVMEDevice.cc]: please add
```c++
probe_queue_cond.notify_all();
```
to notify all waiters.

Comment (tone-zhang) [reply src/os/bluestore/NVMEDevice.cc]: @tchaikov Kefu, thanks for your comments.

File: src/os/bluestore/NVMEDevice.cc [hunk 6]
```diff
@@ -477,7 +477,7 @@ class NVMEManager {
 
  private:
   ceph::mutex lock = ceph::make_mutex("NVMEManager::lock");
-  bool init = false;
+  std::atomic<bool> stopping = false;
```
[the remark this answers is not in the document]
Comment (tone-zhang) [reply src/os/bluestore/NVMEDevice.cc]: Ack. Will update soon. Thanks.

File: src/os/bluestore/NVMEDevice.cc [hunk 7]
```diff
@@ -605,9 +614,12 @@ int NVMEManager::try_get(const spdk_nvme_transport_id& trid, SharedDriverData **
             probe_queue_cond.wait(l);
           }
         }
+        for (auto p : probe_queue) {
+          p->done = true;
+          probe_queue_cond.notify_all();
+        }
```
Comment (tchaikov) [inline src/os/bluestore/NVMEDevice.cc]: @tone-zhang sorry, i missed this. you might need to do the `notify_all()` out of the `for (auto p : probe_queue)` loop.

Comment (tone-zhang) [reply src/os/bluestore/NVMEDevice.cc]: @tchaikov thanks!

**Comment by tchaikov** [comment]
retest this please.

**Comment by tone-zhang** [comment]
@tchaikov Kefu, thanks a lot!

**Pull request merged**